File: docs/en/samples/pre_post_processing/python/langchain/agent.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import asyncio
from datetime import datetime

from langchain.agents import create_agent
from langchain.agents.middleware import wrap_tool_call
from langchain_core.messages import ToolMessage
from langchain_google_vertexai import ChatVertexAI
from toolbox_langchain import ToolboxClient

system_prompt = """
  You're a helpful hotel assistant. You handle hotel searching, booking and
  cancellations. When the user searches for a hotel, mention it's name, id,
  location and price tier. Always mention hotel ids while performing any
  searches. This is very important for any operations. For any bookings or
  cancellations, please provide the appropriate confirmation. Be sure to
  update checkin or checkout dates if mentioned by the user.
  Don't ask for confirmations from the user.
"""


# Pre processing
@wrap_tool_call
async def enforce_business_rules(request, handler):
    """
    Business Logic Validation:
    Enforces max stay duration (e.g., max 14 days).
    """
    tool_call = request.tool_call
    name = tool_call["name"]
    args = tool_call["args"]

    print(f"POLICY CHECK: Intercepting '{name}'")

    if name == "update-hotel":
        if "checkin_date" in args and "checkout_date" in args:
            try:
                start = datetime.fromisoformat(args["checkin_date"])
                end = datetime.fromisoformat(args["checkout_date"])
                duration = (end - start).days

                if duration > 14:
                    print("BLOCKED: Stay too long")
                    return ToolMessage(
                        content="Error: Maximum stay duration is 14 days.",
                        tool_call_id=tool_call["id"],
                    )
            except ValueError:
                pass  # Ignore invalid date formats

    return await handler(request)


# Post processing
@wrap_tool_call
async def enrich_response(request, handler):
    """
    Post-Processing & Enrichment:
    Adds loyalty points information to successful bookings.
    Standardizes output format.
    """
    result = await handler(request)

    if isinstance(result, ToolMessage):
        content = str(result.content)
        tool_name = request.tool_call["name"]

        if tool_name == "book-hotel" and "Error" not in content:
            loyalty_bonus = 500
            result.content = f"Booking Confirmed!\n You earned {loyalty_bonus} Loyalty Points with this stay.\n\nSystem Details: {content}"

    return result


async def main():
    async with ToolboxClient("http://127.0.0.1:5000") as client:
        tools = await client.aload_toolset("my-toolset")
        model = ChatVertexAI(model="gemini-2.5-flash")
        agent = create_agent(
            system_prompt=system_prompt,
            model=model,
            tools=tools,
            middleware=[enforce_business_rules, enrich_response],
        )

        user_input = "Book hotel with id 3."
        response = await agent.ainvoke(
            {"messages": [{"role": "user", "content": user_input}]}
        )

        print("-" * 50)
        print("Final Client Response:")
        last_ai_msg = response["messages"][-1].content
        print(f"AI: {last_ai_msg}")

        # Test Pre-processing
        print("-" * 50)
        user_input = "Update my hotel with id 3 with checkin date 2025-01-18 and checkout date 2025-01-20"
        response = await agent.ainvoke(
            {"messages": [{"role": "user", "content": user_input}]}
        )
        last_ai_msg = response["messages"][-1].content
        print(f"AI: {last_ai_msg}")


if __name__ == "__main__":
    asyncio.run(main())

File: docs/en/samples/pre_post_processing/js/langchain/agent.js
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { ToolboxClient } from "@toolbox-sdk/core";
import { ChatVertexAI } from "@langchain/google-vertexai";
import { AgentExecutor, createToolCallingAgent } from "langchain/agents";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { tool } from "@langchain/core/tools";
import { fileURLToPath } from "url";
import process from "process";

const systemPrompt = `
  You're a helpful hotel assistant. You handle hotel searching, booking and
  cancellations. When the user searches for a hotel, mention it's name, id,
  location and price tier. Always mention hotel ids while performing any
  searches. This is very important for any operations. For any bookings or
  cancellations, please provide the appropriate confirmation. Be sure to
  update checkin or checkout dates if mentioned by the user.
  Don't ask for confirmations from the user.
`;

/**
 * Pre-processing: Enforce Business Rules
 * @param {string} name 
 * @param {object} args 
 * @returns {string|null} Error message if blocked, null otherwise
 */
function checkBusinessRules(name, args) {
  console.log(`POLICY CHECK: Intercepting '${name}'`);

  if (name === "update-hotel" && args.checkin_date && args.checkout_date) {
    try {
      const start = new Date(args.checkin_date);
      const end = new Date(args.checkout_date);
      const duration = (end - start) / (1000 * 60 * 60 * 24); // days

      if (duration > 14) {
        console.log("BLOCKED: Stay too long");
        return "Error: Maximum stay duration is 14 days."; 
      }
    } catch (e) {
      // Ignore invalid dates
    }
  }
  return null;
}

/**
 * Post-processing: Enrich Response
 * @param {string} name 
 * @param {any} result 
 * @returns {any} Enriched result
 */
function enrichResponse(name, result) {
  let content = result;
  if (typeof result === 'object' && result !== null && result.content) {
      content = result.content;
  }

  if (name === "book-hotel" && typeof content === 'string' && !content.includes("Error")) {
      const loyaltyBonus = 500;
      const enrichedContent = `Booking Confirmed!\n You earned ${loyaltyBonus} Loyalty Points with this stay.\n\nSystem Details: ${content}`;
      
      if (typeof result === 'object' && result !== null) {
          result.content = enrichedContent;
          return result;
      }
      return enrichedContent;
  }
  return result;
}

/**
 * Wraps a tool to add pre- and post-processing logic.
 * @param {import("@langchain/core/tools").StructuredTool} toolInstance
 */
function wrapToolWithBusinessLogic(toolInstance) {
  const originalInvoke = toolInstance.invoke.bind(toolInstance);

  toolInstance.invoke = async (input, config) => {
    // 1. Pre-processing
    const validationError = checkBusinessRules(toolInstance.name, input);
    if (validationError) {
      return validationError;
    }

    // 2. Execution
    const result = await originalInvoke(input, config);

    // 3. Post-processing
    return enrichResponse(toolInstance.name, result);
  };

  return toolInstance;
}

/**
 * Helper to run a single turn.
 */
async function runTurn(agentExecutor, input) {
  console.log(`\nUSER: '${input}'`);
  const result = await agentExecutor.invoke({ input });
  console.log("-".repeat(50));
  console.log("Final Client Response:");
  console.log(`AI: ${result.output}`);
}

async function main() {
  const client = new ToolboxClient("http://127.0.0.1:5000");
  
  try {
    const rawTools = await client.loadToolset("my-toolset");
    
    // Convert & Wrap tools
    const tools = rawTools
      .map(t => tool(t, {
          name: t.getName(),
          description: t.getDescription(),
          schema: t.getParamSchema()
      }))
      .map(wrapToolWithBusinessLogic);

    const model = new ChatVertexAI({
      model: "gemini-2.5-flash",
      temperature: 0,
    });

    const prompt = ChatPromptTemplate.fromMessages([
      ["system", systemPrompt],
      ["placeholder", "{chat_history}"],
      ["human", "{input}"],
      ["placeholder", "{agent_scratchpad}"],
    ]);

    const agent = createToolCallingAgent({ llm: model, tools, prompt });
    const agentExecutor = new AgentExecutor({ agent, tools, verbose: false });

    // Turn 1: Booking
    await runTurn(agentExecutor, "Book hotel with id 3.");
    
    // Turn 2: Policy Violation
    await runTurn(agentExecutor, "Update my hotel with id 3 with checkin date 2025-01-18 and checkout date 2025-02-10");

  } catch (error) {
    console.error("Error running agent:", error);
  }
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main();
}

export { main };